### PR TITLE
Sharing

### PR DESCRIPTION
**Pull request by cody-constine-ttd**

### CODE REVIEW

File: lib/keycontainer.h [hunk 1]
```diff
@@ -13,12 +13,22 @@ namespace uid2
 	public:
 		KeyContainer() = default;
 
+        KeyContainer(int _callerSiteId, int _masterKeysetId, int _defaultKeysetId, long _tokenExpirySeconds)
```
Comment (zaiweidu) [inline lib/keycontainer.h:15]: why are indentations different? It seems you need to re-format all files.

File: lib/keycontainer.h [hunk 2]
```diff
@@ -55,15 +66,78 @@ namespace uid2
...
+        const Key* getDefaultKey(Timestamp now) const {
+            return GetActiveKeysetKey(defaultKeySetId, now);
+        }
+
+
+        int getTokenExpirySeconds() const {
+            return tokenExpirySeconds;
+        }
+
+        void setTokenExpirySeconds(long tokenExpirySeconds) {
```
Comment (zaiweidu) [inline lib/keycontainer.h]: **long** is at least the same size as **int**. Its actual width is determined by compiler & architecture. You may use **int64_t**. **long long** is identical to int64_t by implicit conversion.

File: lib/keyparser.cpp [hunk 3]
```diff
@@ -32,14 +39,48 @@ namespace uid2
...
+        int masterKeysetId;
+        if(!ExtractInt(*body, MASTER_KEYSET_ID_NAME, masterKeysetId)) {
+            out_err = "returned json does not contain a master keyset id";
+            return false;
+        }
+        out_container.setMasterKeySetId(masterKeysetId);
+
+        int defaultKeysetId;
+        if(!ExtractInt(*body, DEFAULT_KEYSET_ID_NAME, defaultKeysetId)) {
+            defaultKeysetId = -1;
```
Comment (zaiweidu) [inline lib/keyparser.cpp]: If use -1 to represent null int, I'd have a macro defined for _null Int32_.

File: lib/uid2_token_generator.cpp [hunk 4]
```diff
@@ -1,4 +1,4 @@
-#include "uid2_token_generator.h"
+#include "uid2/uid2_token_generator.h"
```
Comment (zaiweidu) [inline lib/uid2_token_generator.cpp]: Can we keep the same format of header file reference? "" and <> are mixed here.

File: include/uid2/timestamp.h [hunk 5]
```diff
@@ -18,7 +18,7 @@ namespace uid2
 		std::int64_t GetEpochMilli() const { return EpochMilli; }
 		bool IsZero() const { return EpochMilli == 0; }
 
-		Timestamp AddSeconds(int seconds) const { return Timestamp(EpochMilli + seconds * 1000); }
+		Timestamp AddSeconds(std::int64_t seconds) const { return Timestamp(EpochMilli + seconds * 1000); }
```
Comment (jon8787) [inline include/uid2/timestamp.h:21]: curious, why this change?

File: include/uid2/uid2client.h [hunk 6]
```diff
@@ -15,7 +15,9 @@ namespace uid2
 		virtual ~IUID2Client() = default;
 
 		virtual RefreshResult Refresh() = 0;
+        virtual RefreshResult RefreshJson(const std::string& json) = 0;
```
Comment (jon8787) [inline include/uid2/uid2client.h]: can we remove this, as it should only be exposed internally

File: include/uid2/uid2client.h [hunk 7]
```diff
@@ -15,7 +15,9 @@ namespace uid2
...
 		virtual DecryptionResult Decrypt(const std::string& token, Timestamp now) = 0;
+        virtual EncryptionResult Encrypt(const std::string& uid, Timestamp now) = 0;
```
Comment (jon8787) [inline include/uid2/uid2client.h]: likewise with Timestamp, should only be internal

File: lib/key.h [hunk 8]
```diff
@@ -4,13 +4,15 @@
 
 #include <cstdint>
 #include <vector>
+#include <optional>
```
Comment (jon8787) [inline lib/key.h]: this should be removed, curious why did you decide not to use it?

File: lib/uid2client.cpp [hunk 9]
```diff
@@ -36,8 +37,10 @@ namespace uid2
 				// TODO: non-https endpoint warning
 			}
 
+            // TODO: Figure out how to get sdk version
```
Comment (jon8787) [inline lib/uid2client.cpp]: for now, can we add a hardcoded version?

File: test/encryption_tests_v2.cpp [hunk 10]
```diff
@@ -2,7 +2,7 @@
 
 #include "base64.h"
 #include "key.h"
-#include "uid2_token_generator.h"
+#include "uid2/uid2_token_generator.h"
```
Comment (jon8787) [inline test/encryption_tests_v2.cpp:5]: can we move this to the cpp?

File: test/encryption_tests_v4.cpp [hunk 11]
```diff
@@ -26,10 +31,13 @@ static const int SITE_ID2 = 2;
 static const std::uint8_t MASTER_SECRET[] = { 139, 37, 241, 173, 18, 92, 36, 232, 165, 168, 23, 18, 38, 195, 123, 92, 160, 136, 185, 40, 91, 173, 165, 221, 168, 16, 169, 164, 38, 139, 8, 155 };
 static const std::uint8_t SITE_SECRET[] = { 32, 251, 7, 194, 132, 154, 250, 86, 202, 116, 104, 29, 131, 192, 139, 215, 48, 164, 11, 65, 226, 110, 167, 14, 108, 51, 254, 125, 65, 24, 23, 133 };
 static const Timestamp NOW = Timestamp::Now();
-static const Key MASTER_KEY{MASTER_KEY_ID, -1, NOW.AddDays(-1), NOW, NOW.AddDays(1), GetMasterSecret()};
-static const Key SITE_KEY{SITE_KEY_ID, SITE_ID, NOW.AddDays(-10), NOW.AddDays(-9), NOW.AddDays(1), GetSiteSecret()};
+static const Key MASTER_KEY{MASTER_KEY_ID, -1, 1, NOW.AddDays(-1), NOW, NOW.AddDays(1), GetMasterSecret()};
+static const Key SITE_KEY{SITE_KEY_ID, SITE_ID, 99999, NOW.AddDays(-10), NOW.AddDays(-9), NOW.AddDays(1), GetSiteSecret()};
```
Comment (jon8787) [inline test/encryption_tests_v4.cpp]: should we be using DEFAULT_KEYSET_ID here and MASTER_KEYSET_ID above?

File: test/encryption_tests_v4.cpp [hunk 12]
```diff
@@ -240,6 +248,237 @@ TEST(EncryptDataTestsV4, RawUidProducesCorrectIdentityTypeInToken)
...
+    EXPECT_EQ(DecryptionStatus::SUCCESS, res.GetStatus());
+    EXPECT_EQ(EXAMPLE_UID, res.GetUid());
+}
+
+TEST(SharingTests, SharingTokenIsV4)
+{
+    auto clientAndToken = SharingSetupAndEncrypt();
+    auto advertisingToken = clientAndToken.advertisingToken;
+
+    bool containsBase64SepcialChars = advertisingToken.find("+") == std::string::npos || advertisingToken.find("/") == std::string::npos || advertisingToken.find("=") == std::string::npos;
```
Comment (jon8787) [inline test/encryption_tests_v4.cpp]: typo

File: lib/key.h [hunk 13]
```diff
@@ -4,13 +4,15 @@
...
 
 namespace uid2
 {
 	struct Key
 	{
 		std::int64_t id;
 		int siteId;
+        int keysetId;
```
Comment (jon8787) [inline lib/key.h:15]: we can initialize these to default values, eg "int keysetId = -1" or 0

also on this topic, if we're using -1 as a magic value, we can put "const int NO_KEYSET = -1" and use NO_KEYSET everywhere instead of -1

File: include/uid2/timestamp.h (hunk 5 above)
Comment (cody-constine-ttd) [reply include/uid2/timestamp.h:21]: It was overflowing when adding the token expiry

File: include/uid2/uid2client.h (hunk 6 above)
Comment (cody-constine-ttd) [reply include/uid2/uid2client.h]: I need it exposed in the interface, so I can use it in the tests

File: lib/key.h (hunk 8 above)
Comment (cody-constine-ttd) [reply lib/key.h]: O oops I was trying something but it was a c++17 feature

File: include/uid2/uid2client.h (hunk 6 above)
Comment (jon8787) [reply include/uid2/uid2client.h]: Yes, take a look at the c# and java sdks, we should be able to do similar here

File: test/encryption_tests_v2.cpp (hunk 10 above)
Comment (cody-constine-ttd) [reply test/encryption_tests_v2.cpp:5]: It is in the cpp?

Comment (jon8787) [reply test/encryption_tests_v2.cpp:5]: no idea what I was thinking

**Review by zaiweidu — APPROVED** [review]
LGTM
